### PR TITLE
Accept non-unit type annotations in @quantity_input

### LINKED ISSUE #7661: Make @quantity_input pass over arguments with non-unit annotations
The `@astropy.units.quantity_input` decorator currently raises an exception if an argument has an annotation that cannot be converted to a unit.  This has the side effect that type hint annotations and annotations used by different decorators cannot be used in functions decorated by `@quantity_input`.

Here is a minimal example (using Astropy 3.0.2):
```Python
>>> import astropy.units as u

>>> @u.quantity_input
... def f(x: u.m, y: str) -> u.m:
...    return x

>>> f(1 * u.m, 'x')
TypeError: <class 'str'> can not be converted to a Unit
```

My suggestion would be for `@quantity_input` to skip arguments with annotations that are not units or cannot be converted to units.  However, if the annotation is a string that cannot be converted to a unit, then I would propose raising a `TypeError` (or maybe a `u.UnitsError` or a `u.UnitConversionError`) since there could be a typo in a string that could have been intended to be a unit.

Thanks!
-Nick

### PR DESCRIPTION
**Pull request by ritiek**
Closes #7661.

Instead of raising an error, `@quantity_input` will skip any non-unit type annotations in function parameters. So, code such as below won't result in an error:

```python
>>> import astropy.units as u

>>> @u.quantity_input
... def f(x: u.m, y: str) -> u.m:
...    return x

>>> f(1 * u.m, 'x')
<Quantity 1. m>
# this would otherwise give
# TypeError: <class 'str'> can not be converted to a Unit
```

It will however, raise a `ValueError` in cases like these:
```python
>>> import astropy.units as u

>>> @u.quantity_input
... # consider the typo "kilogrammm" instead of "kilogram"
... def f(x: u.m, y: "kilogrammm") -> u.m:
...    return x

>>> f(1 * u.m, 2 * u.kg)
ValueError: Invalid unit or physical type 'kilogrammm'.
```

TLDR; @quantity_input now ignores any parameter in function definition which is non-unit type annotated, but will raise a `ValueError` if a string is passed and `Unit(string)` is unable to parse it (like above example).

### CODE REVIEW

**Comment by ritiek** [comment]
cc @namurphy.

File: astropy/units/decorators.py [hunk 1]
```diff
@@ -50,8 +52,12 @@ def _validate_arg_value(param_name, func_name, arg, targets, equivalencies):
 
     for allowed_unit in allowed_units:
         try:
-            is_equivalent = arg.unit.is_equivalent(allowed_unit,
-                                                   equivalencies=equivalencies)
+            # any other annotation was provided except for ``astropy.unit``
+            if inspect.isclass(allowed_unit):
```
Comment (Cadair) [inline astropy/units/decorators.py]: Can you explain this to me? I don't understand how it helps to filter out `Unit` classes.

**Comment by bsipocz** [comment]
@ritiek - any plans to address the review in the coming 2 days to get this into the new release, or you prefer to get this remilestoned to 4.0?

**Comment by ritiek** [comment]
@bsipocz My apologies, I completely forgot about this PR.

I'm currently in the midst of my final exams and probably won't have time to work on this until this weekend. It would probably be best if you guys could re-milestone as I wouldn't want the release to delay cuz of this PR.

**Comment by bsipocz** [comment]
@ritiek - thanks for getting back, and good luck for the exams.

**Comment by maciekgroch** [comment]
I hope you passed your exams :). Is there any plan to close the issue?

Thanks for help.

**Comment by ritiek** [comment]
#8984 supersedes this.
